tests/test_collection.py

```diff
@@ -0,0 +1,366 @@
+        ):
+            self.assertEqual(schema.dump(schema.load(data_in)), data_in)
+
+    def test_sequence_of_optional_dataclass(self):
+        @dataclass
+        class Elm:
+            value: int
+
+        @dataclass
+        class SequenceIntSequence:
+            value: Sequence[Optional[Elm]]
+
+        schema = SequenceIntSequence.Schema()
+        for data_in in (
+            {"value": []},
+            {"value": [None]},
+            {"value": [{"value": 1}]},
+        ):
+            self.assertEqual(schema.dump(schema.load(data_in)), data_in)
+
+    def test_serializing_sets_as_sequence_is_possible_but_risky(self):
+        @dataclass
+        class StrSequence:
+            value: Sequence[str]
+
+        str_schema = StrSequence.Schema()
+
+        # you can dump it, but order will be random
+        sequence = (
+            "this",
+            "has",
+            "only",
+            "1",
+            "chance",
+            "in",
+            "40M",
+            "to",
+            "keep",
+            "it's",
+            "ordered",
+        )
+        dump = str_schema.dump(StrSequence(value=set(sequence)))
+        self.assertNotEqual(
+            dump,
+            {"value": sequence},
+        )
+
+        @dataclass
+        class IntSequence:
+            value: Sequence[int]
+
+        int_schema = IntSequence.Schema()
+
+        # and to make it worse, int sets are a valid sequence that keep order....
+        dump = int_schema.dump(IntSequence(value={1, 2, 3, 4, 5}))
+        self.assertEqual(dump, {"value": [1, 2, 3, 4, 5]})
+
+
+class TestSetField(unittest.TestCase):
+    def test_simple(self):
+        @dataclass
+        class IntSet:
+            value: Set[int]
+
+        schema = IntSet.Schema()
+        data_in = {"value": list(range(5))}
+        loaded = schema.load(data_in)
+        self.assertEqual(loaded, IntSet(value=set(range(5))))
+        self.assertCountEqual(schema.dump(loaded)["value"], data_in["value"])
+
+    def test_set_no_arg(self):
+        @dataclass
+        class AnySet:
+            value: Set
+
+        schema = AnySet.Schema()
+
+        # can load a set of mixed kind
+        data_in = {"value": [1, "2", 3.0, (1, 2, 3)]}
+        loaded = schema.load(data_in)
+        self.assertEqual(loaded, AnySet(value={1, "2", 3.0, (1, 2, 3)}))
+        self.assertCountEqual(schema.dump(loaded)["value"], data_in["value"])
+
+    def test_optional_set(self):
+        @dataclass
+        class IntOptionalSet:
+            value: Optional[Set[int]]
+
+        schema = IntOptionalSet.Schema()
+        for data_in in ({"value": list(range(5))}, {"value": []}, {"value": None}):
+            dumped_value = schema.dump(schema.load(data_in))["value"]
+            value = schema.dump(schema.load(data_in))["value"]
+            if value is None:
+                self.assertIsNone(dumped_value)
+            else:
+                self.assertCountEqual(dumped_value, value)
```

Comment:
What are you trying to test here ?
Why computing `schema.dump(schema.load(data_in))["value"]` twice ?

Reply:
Yep, overly complicated test that missed the point. I just wanted to test the None case here I guess. I will simplify.